working/0110-incremental-sound-nnbd/feature-specification.md

```diff
@@ -0,0 +1,374 @@
+The same is true for `{ int ? - 3 : 3 }` if we allow this.
+
+
+## Static semantics
+
+
+### Static errors
+
+We say that a type `T` is **nullable** if `Null <: T`.  This is equivalent to
+the syntactic criterion that `T` is any of:
+  - `Null`
+  - `S?` for some `S`
+  - `FutureOr<S>` for some `S` where `S` is nullable
+  - `dynamic`
+  - `void`
+
+We say that a type `T` is **non-nullable** if `T <: Object`.  This is equivalent
+to the syntactic criterion that `T` is any of:
+  - `Object`, `int`, `bool`, `Never`, `Function`
+  - Any function type
+  - Any class type or generic class type
+  - `FutureOr<S>` where `S` is non-nullable
+  - `X extends S` where `S` is non-nullable
+  - `X & S` where `S` is non-nullable
+
+Note that there are types which are neither nullable nor non-nullable.  For
+example `X extends T` where `T` is nullable is neither nullable nor
+non-nullable.
+
+We say that a type `T` is **potentially nullable** if `T` is not non-nullable.
+Note that this is different from saying that `T` is nullable.  For example, a
+type variable `X extends Object?` is a type which is potentially nullable but
+not nullable.
+
+We say that a type `T` is **potentially non-nullable** if `T` is not nullable.
+Note that this is different from saying that `T` is non-nullable.  For example,
+a type variable `X extends Object?` is a type which is potentially non-nullable
+but not non-nullable.
+
+It is an error to call a method, setter, getter or operator on an expression
+whose type is potentially nullable and not `dynamic`, except for the methods,
+setters, getters, and operators on `Object`.
+
+It is an error to read a field or tear off a method from an expression whose
+type is potentially nullable and not `dynamic`, except for the methods and
+fields on `Object`.
+
+It is an error to call an expression whose type is potentially nullable and not
+`dynamic`.
+
+It is an error for an instance field with no static initializer to have a
+potentially nullable type unless the variable or field is marked with the `lazy`
+modifier.
+
+It is an error if a local variable that is potentially non-nullable and is not
+marked `lazy` is used before it is definitely assigned (see Definite Assignment
+below).
+
+It is an error if a method, function, getter, or closure with a potentially
+non-nullable return type does not definitely complete (see Definite Completion
+below).
+
+It is an error if an optional parameter (named or otherwise) with no default
+value has a potentially non-nullable type.
+
+It is an error if a required named parameter has a default value.
+
+It is an error if a named parameter that is part of a `required` group is not
+bound to an argument at a call site.
+
+It is an error to call the default `List` constructor with a length argument and
+a type argument which is potentially non-nullable.
+
+For the purposes of errors and warnings, the null aware operators `?.` and `?..`
+are checked as if the receiver of the operator had non-nullable type.
```

Comment:
Should it be an error to use `?.` or `?..` on a non-nullable type?

Reply:
Not sure.  Maybe?  It's nice to get some warning at least, but it makes changing a type from nullable to non-nullable more of a breaking change.  e.g. in

```
  library lib;
  int ? x;
  int? foo() => 3
```

you could change both `int?`s to `int` with no downstream breakage unless we make it an error to do `foo()?.isEven` if `foo` returns int.

Reply:
Ok, I've added the following:

```
It is a warning to use a null aware operator (`?.`, `?..`, `??`, `??=`, or
`...?`) on a non-nullable receiver.
```